src/main/java/spoon/reflect/visitor/DefaultJavaPrettyPrinter.java

```diff
@@ -815,9 +815,9 @@ public void visitCtDo(CtDo doLoop) {
 							target.setImplicit(false);
 						}
 					}
-					printer.getPrinterHelper().snapshotLength();
 					scan(target);
-					if (printer.getPrinterHelper().hasNewContent()) {
+					// the implicit drives the separator
+					if (!target.isImplicit()) {
 						printer.writeSeparator(".");
```

Comment:
move scan call above into this if block?
Then it is clear when it produces output.